EIPS/eip-7709.md

```diff
@@ -0,0 +1,138 @@
+---
+eip: 7709
+title: Read BLOCKHASH from storage and update cost
+description: Read the `BLOCKHASH (0x40)` opcode from the EIP-2935 system contract storage and adjust its gas cost to reflect storage access.
+author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech), Tanishq Jasoria (@tanishqjasoria), Ignacio Hagopian (@jsign), Jochem Brouwer (@jochem-brouwer), Gabriel Rocheleau (@gabrocheleau)
+discussions-to: https://ethereum-magicians.org/t/eip-7709-read-blockhash-opcode-from-storage-and-adjust-gas-cost/20052
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-05-18
+requires: 2935, 6800
+---
+
+## Abstract
+
+Update the `BLOCKHASH (0x40)` opcode to read and serve from the system contract storage to allow for stateless execution. Update the `BLOCKHASH (0x40)` opcode gas cost to reflect storage access cost, and include storage accesses in the Verkle block witness.
+
+## Motivation
+
+The `BLOCKHASH (0x40)` opcode currently assumes that the client has knowledge of the previous blocks, which in Verkle, would prevent stateless execution. This EIP assumes that [EIP-2935](./eip-2935.md) has been implemented, and that blockhashes can be retrieved from the systems contract storage. By updating the behavior of `BLOCKHASH (0x40)` to directly read and serve from state through the system contract storage, we allow Verkle blocks to include a storage access witness, which will allow stateless execution.
+
+The motivation behind the updated gas cost is to match the real cost of the operation, which is equivalent to an `SLOAD`.
+
+## Specification
+
+| Parameter                 | Value                                        |
+| ------------------------- | -------------------------------------------- |
+| `FORK_TIMESTAMP`          | TBD                                          |
+| `HISTORY_STORAGE_ADDRESS` | `0x25a219378dad9b3503c8268c9ca836a52427a4fb` |
```

Comment:
I would set this to `TBD` for now, as the code for the contract needs to be finalized and deployed with eip 2935. This part will not change with 7709, as we need to use the same contract for continuity.

Reply:
updated